perception/sandbox/polygrasp/src/polygrasp/robot_interface.py

```diff
@@ -0,0 +1,151 @@
+
+    des_ori = plane_rot * R.from_euler("x", 90, degrees=True) * R.from_euler("y", 90, degrees=True)
+    des_ori_quat = des_ori.as_quat()
+
+    return grasp_point, grasp_approach_delta, des_ori_quat
+
+
+def grasp_to_pose(grasp: graspnetAPI.Grasp):
+    return grasp.translation, R.from_matrix(grasp.rotation_matrix).as_quat()
+
+
+def compute_quat_dist(a, b):
+    return torch.acos((2 * (a * b).sum() ** 2 - 1).clip(-1, 1))
+
+
+def min_dist_grasp(default_ee_pose, grasps):
+    with torch.no_grad():
+        rots_as_quat = [
+            torch.Tensor(R.from_matrix(grasp.rotation_matrix).as_quat()) for grasp in grasps
+        ]
+        dists = [compute_quat_dist(rot, default_ee_pose) for rot in rots_as_quat]
+        i = torch.argmin(torch.Tensor(dists)).item()
+    log.info(f"Grasp {i} has angle {dists[i]} from reference.")
+    return grasps[i], i
+
+
+class GraspingRobotInterface(polymetis.RobotInterface):
+    def __init__(
+        self, gripper: polymetis.GripperInterface, k_approach=1.5, k_grasp=0.72, *args, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.gripper = hydra.utils.instantiate(gripper)
+
+        self.default_ee_pose = torch.Tensor([0.9418, 0.3289, -0.0368, -0.0592])
+        self.k_approach = k_approach
+        self.k_grasp = k_grasp
+
+    def gripper_open(self):
+        self.gripper.goto(1, 1, 1)
+
+    def gripper_close(self):
+        self.gripper.goto(0, 1, 1)
+
+    def move_until_success(
+        self, position, orientation, time_to_go, max_attempts=5, success_dist=0.2
+    ):
+        states = []
+        for _ in range(max_attempts):
+            states += self.move_to_ee_pose(
+                position=position, orientation=orientation, time_to_go=time_to_go
+            )
+            curr_ee_pos, curr_ee_ori = self.get_ee_pose()
+
+            log.info(f"Dist to goal: {torch.linalg.norm(curr_ee_pos - position)}")
+
+            if (
+                states[-1].prev_command_successful
+                and torch.linalg.norm(curr_ee_pos - position) < success_dist
+            ):  # TODO: orientation diff
+                break
+        return states
+
+    def select_grasp(self, grasps: graspnetAPI.GraspGroup) -> graspnetAPI.Grasp:
+        with torch.no_grad():
+            feasible_i = []
+            for i, grasp in enumerate(grasps):
+                grasp_point, grasp_approach_delta, des_ori_quat = compute_des_pose(grasp)
+                point_a = grasp_point + self.k_approach * grasp_approach_delta
+                point_b = grasp_point + self.k_grasp * grasp_approach_delta
+
+                def check_feasibility(point):
+                    ik_sol = self.robot_model.inverse_kinematics(
+                        torch.Tensor(point), torch.Tensor(des_ori_quat)
+                    )
+                    ee_pos, ee_quat = self.robot_model.forward_kinematics(ik_sol)
+                    return torch.linalg.norm(ee_pos - point) < 0.001
```

Comment:
Should also check for orientation here? Depends on how the IK cost is tuned, you could have near zero position error but a large orientation error at times.